src/common/buffer.cc

```diff
@@ -1520,7 +1536,16 @@ using namespace ceph;
     const unsigned free_in_last = get_append_buffer_unused_tail_length();
     const unsigned first_round = std::min(len, free_in_last);
     if (first_round) {
-      _buffers.back().append(data, first_round);
+      // _buffers and carriage can desynchronize when 1) a new ptr
+      // we don't own has been added into the _buffers 2) _buffers
+      // has been emptied as as a result of std::move or stolen by
+      // claim_append.
+      if (unlikely(_carriage != &_buffers.back())) {
+        auto bptr = ptr_node::create(*_carriage, _carriage->length(), 0);
+	_carriage = bptr.get();
+	_buffers.push_back(*bptr.release());
+      }
+      _carriage->append(data, first_round);
```

Comment:
I'm not following what this patch is trying to improve.  Is the motivation that _buffers.back() is slower than using _carriage directly or something?  It seems like we don't get away from _buffers.back() though because we have to check whether it's != _buffers.back() here anyway

Reply:
The idea behind https://github.com/ceph/ceph/pull/25077/commits/7e711ae275d2315f35874769ee6a97700fbe9a2c introducing `_carriage` is to prevent memory waste due to mixing with e.g. `::append(const bufferlist&&)`. This would happen when:
```cpp
bufferlist bl;
bl.append("abc", 3); // allocates 4 KB
bl.append(other_nonempty_bl); // shallow copy enlarges bl._buffers; bl._buffers.back() is NOT writable for us
bl.append("def", 3); // thanks to _carriage we can reuse the initially allocated 4 KB
```

Reply:
oh i see